ci/cloudbuild/notifiers/alerts/function/function.cc

```diff
@@ -0,0 +1,95 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <google/cloud/functions/cloud_event.h>
+#include <cppcodec/base64_rfc4648.hpp>
+#include <curl/curl.h>
+#include <fmt/core.h>
+#include <nlohmann/json.hpp>
+#include <iostream>
+#include <memory>
+#include <stdexcept>
+
+namespace {
+
+struct BuildStatus {
+  nlohmann::json build;
+  std::string status;
+};
+
+// Parses the Pub/Sub message within the given `event`, and returns the build
+// status and the embedded Build object from GCB. See also
+// https://cloud.google.com/build/docs/api/reference/rest/v1/projects.builds#Build
+BuildStatus ParseBuildStatus(google::cloud::functions::CloudEvent event) {
+  if (event.data_content_type().value_or("") != "application/json") {
+    throw std::runtime_error("expected application/json data");
+  }
+  auto const payload = nlohmann::json::parse(event.data().value_or("{}"));
+  if (payload.count("message") == 0) {
+    throw std::runtime_error("missing embedded Pub/Sub message");
+  }
+  auto const pubsub = payload["message"];
+  if (pubsub.count("attributes") == 0 || pubsub.count("data") == 0) {
+    throw std::runtime_error("missing Pub/Sub attributes or data");
+  }
+  auto const data = cppcodec::base64_rfc4648::decode<std::string>(
+      pubsub["data"].get<std::string>());
+  return BuildStatus{nlohmann::json::parse(data),
+                     pubsub["attributes"].value("status", "")};
+}
+
+nlohmann::json MakeChatPayload(BuildStatus const& bs) {
+  auto const trigger_name = bs.build["substitutions"].value("TRIGGER_NAME", "");
+  auto const log_url = bs.build.value("logUrl", "");
+  auto text = fmt::format(R""(Build failed: *{trigger_name}* {log_url})"",
```

Comment:
nit: for a short format like this, where the arguments are not repeated, you could have used positional arguments:

```cc
auto text = fmt::format("Build failed: *{}* {}", trigger_name, log_url);
```

Reply:
Good point. The format was longer originally and it was simplified over time. This is better -- and one line.

Also, I just noticed https://xkcd.com/1296/ -- HAHA super relevant here.